private/pkg/git/object_reader.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/bufbuild/buf/private/pkg/command"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"go.uber.org/multierr"
+)
+
+// exitTime is the amount of time we'll wait for git-cat-file(1) to exit.
+var exitTime = 5 * time.Second
+
+type objectReader struct {
+	rx      *bufio.Reader
+	tx      io.WriteCloser
+	process command.Process
+}
+
+func newObjectReader(gitDirPath string, runner command.Runner) (*objectReader, error) {
+	gitDirPath = normalpath.Unnormalize(gitDirPath)
+	if err := validateDirPathExists(gitDirPath); err != nil {
+		return nil, err
+	}
+	gitDirPath, err := filepath.Abs(gitDirPath)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: do we need to validate `gitDirPath` is not jumping context?
```

Comment:
Had an offline conversation, this is unlikely to be a problem in practice. I think there is merit to having a validation code path, but I would not block my review on this.